Zend/zend_inheritance.c

```diff
@@ -2174,7 +2204,92 @@ static void zend_do_traits_method_binding(zend_class_entry *ce, zend_class_entry
 }
 /* }}} */
 
-static zend_class_entry* find_first_definition(zend_class_entry *ce, zend_class_entry **traits, size_t current_trait, zend_string *prop_name, zend_class_entry *colliding_ce) /* {{{ */
+static zend_class_entry* find_first_constant_definition(zend_class_entry *ce, zend_class_entry **traits, size_t current_trait, zend_string *constant_name, zend_class_entry *colliding_ce) /* {{{ */
+{
+	size_t i;
+
+	if (colliding_ce == ce) {
+		for (i = 0; i < current_trait; i++) {
+			if (traits[i]
+				&& zend_hash_exists(&traits[i]->constants_table, constant_name)) {
+				return traits[i];
+			}
+		}
+	}
+
+	return colliding_ce;
+}
+/* }}} */
+
+static bool do_trait_constant_check(zend_class_entry *ce, zend_class_constant *trait_constant, zend_string *name, zend_class_entry **traits, size_t current_trait) /* {{{ */
+{
+	uint32_t flags_mask = ZEND_ACC_PPP_MASK | ZEND_ACC_FINAL;
+
+	zval *zv = zend_hash_find_known_hash(&ce->constants_table, name);
+	if (zv == NULL) {
+		// No existing constant of the same name, so this one can be added
+		return true;
+	}
+
+	zend_class_constant *existing_constant = Z_PTR_P(zv);
+
+	if ((ZEND_CLASS_CONST_FLAGS(trait_constant) & flags_mask) == (ZEND_CLASS_CONST_FLAGS(existing_constant) & flags_mask)) {
+		if (check_trait_property_or_constant_value_compatibility(ce, &trait_constant->value, &existing_constant->value)) {
+			// There is an existing constant which is compatible with the new one, so no need to add it
+			return false;
+		}
+	}
+
+	// There is an existing constant of the same name, and it conflicts with the new one, so let's throw a fatal error
+	zend_error_noreturn(E_COMPILE_ERROR,
+		"%s and %s define the same constant (%s) in the composition of %s. However, the definition differs and is considered incompatible. Class was composed",
+		ZSTR_VAL(find_first_constant_definition(ce, traits, current_trait, name, existing_constant->ce)->name),
+		ZSTR_VAL(trait_constant->ce->name),
+		ZSTR_VAL(name),
+		ZSTR_VAL(ce->name));
+}
+/* }}} */
```

Comment:
I'm confused here, doesn't the compiler complain that no value is returned here (in case of an error)?

And to be sure, ``true`` means "add constant to the hashtable", ``false`` do not, and otherwise error?

Reply:
> doesn't the compiler complain that no value is returned here

The `zend_error_noreturn()` is marked as `ZEND_NORETURN`, so the compiler doesn't complain about it.
But yeah, I'm not sure if it's OK or not, since php-src only does the same thing in a few places.

> And to be sure, true means "add constant to the hashtable", false do not, and otherwise error?

Yes!

The current form clearly states without redundancy that no value is returned to the caller in this condition, but I am not strongly attached to this part.

Would it be clearer to invert the condition and put the error handling in the if-block so that it returns false on the face of the code, or add an explicit abort in the if-block?

Reply:
I think the current order is fine, I was just surprised that the compiler didn't complain about not returning in case of an error, and I didn't know how you would handle it with just a boolean (but then you could still reutrn ``false`` as an error should not add the constant anyway)

Reply:
I've changed the order, as I'm starting to feel that there should be fewer surprises in minor areas.